### PR TITLE
Async Activity Completion

### PR DESCRIPTION
**Pull request by angelazhou32**
This is the Async Activity Completion chapter (chapter 6) for Interacting with Workflows in TypeScript course.

### CODE REVIEW

File: asynchronous-activity-completion/what-is-a-heartbeat.md [hunk 1]
```diff
@@ -0,0 +1,22 @@
+# What is an Activity Heartbeat?
```
Comment (angelazhou32) [inline asynchronous-activity-completion/what-is-a-heartbeat.md:1]: This will be covered in more depth in the Errors course.

File: asynchronous-activity-completion/how-to-async-complete-activities.md [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+  // does some work...
+  await client.complete(taskToken, "Job's done!");
+}
+```
+
+Let's look at the above code step by step:
+
+1. We began by extracting the Task Token that uniquely identifies the current Activity execution.
+2. We then scheduled `doSomeWork` to be executed after a delay. The delay represents a stand-in for more complex asynchronous operations, like waiting for an external API response or processing that happens in a separate system.
+3. We then throw the `CompleteAsyncError` to make the Worker forget about this Activity. This signals that the Activity has not completed yet and is waiting for an external completion, not immediately upon the function's return.
```
Comment (MasonEgger) [inline asynchronous-activity-completion/how-to-async-complete-activities.md:36]:
```suggestion
3. We then threw the `CompleteAsyncError` to make the Worker forget about this Activity. This signaled that the Activity has not completed yet and is waiting for an external completion, not immediately upon the function's return.
```
Past tense to match above

File: asynchronous-activity-completion/how-to-async-complete-activities.md [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+4. Finally, `AsyncCompletionClient` is used to asynchronously communicate the result of the asynchronous Activity back (identified by the Task Token) to the Workflow. The result communicated back in this instance is "Job's done!".
```
Comment (MasonEgger) [inline asynchronous-activity-completion/how-to-async-complete-activities.md:37]:
```suggestion
4. Finally, `AsyncCompletionClient` was used to asynchronously communicate the result of the asynchronous Activity back (identified by the Task Token) to the Workflow. The result communicated back in this instance is "Job's done!".
```

**Review by MasonEgger — CHANGES_REQUESTED** [review]

File: asynchronous-activity-completion/async-completion-vs-signals.md [hunk 4]
```diff
@@ -0,0 +1,48 @@
+# Deciding Between Asynchronous Completion and Signals
+
+When integrating external systems into Temporal Workflows, choosing between Asynchronous Completion and Signals significantly impacts how the Workflow interacts with external processes. The decision between using Asynchronous Completion and Signals involves a few factors.
+
+Asynchronous Completion is best suited for scenarios where the external system's interaction is less predictable or reliable.
```
[the remark this answers is not in the document]
Comment (angelazhou32) [reply asynchronous-activity-completion/async-completion-vs-signals.md:5]: Moved it to below the ideal cases for async completion bullets

File: asynchronous-activity-completion/async-completion-vs-signals.md [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+
+Manual activities, such as a user clicking a link or entering data, occur at an unpredictable pace. These are dependent on the user's actions, which can vary greatly in timing. Signals are designed for scenarios where the Workflow needs to respond to external, asynchronous events. When a user performs an action, a Signal is sent to the Workflow, notifying it of the event without the Workflow having to poll or wait actively for the action. With Signals, a developer can use a very short running Activity (with a very brief `StartToCloseTimeout`) that simply hits a third-party service and then waits for the Activity to complete. Once the Activity completes, a Signal will notify the application that it is ready for the next step.
```
[the remark this answers is not in the document]
Comment (angelazhou32) [reply asynchronous-activity-completion/async-completion-vs-signals.md:46]: fixed!

File: asynchronous-activity-completion/how-to-async-complete-activities.md [hunk 6]
```diff
@@ -0,0 +1,41 @@
+# How to Asynchronously Complete Activities
+
+Asynchronous Activity Completion calls can take a Task Token in as an argument. A Task Token is a unique identifier for an Activity Task Execution. It's essential for Temporal to track the specific Activity instance across machines.
```
[the remark this answers is not in the document]
Comment (angelazhou32) [reply asynchronous-activity-completion/how-to-async-complete-activities.md:3]: fixed

File: asynchronous-activity-completion/how-to-async-complete-activities.md [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+
+1. **Provide Identifying Information**: The Activity provides the external system with identifying information needed to complete the Activity Execution. Identifying information can be a Task Token or a combination of Namespace, Workspace ID, and Activity ID.
+
+2. **Async Activity Completion**: The Activity Function completes in a way that identifies it as waiting to be completed by an external system. You can throw [`CompleteAsyncError`](https://typescript.temporal.io/api/classes/activity.CompleteAsyncError) to make the Worker forget about this Activity. The Activity can then be completed asynchronously elsewhere.
+
+3. **Use Temporal Client**: Optionally, while the external work is being done, the Temporal Client can send heartbeats to update the progress of the Activity Execution. Eventually, the Temporal Client is used to complete the Activity, signaling the end of its execution with the result of the external work. To asynchronously complete an Activity, call[`AsyncCompletionClient`](https://typescript.temporal.io/api/classes/client.AsyncCompletionClient) which is a client for asynchronous completion and heartbeating of Activities.
```
[the remark this answers is not in the document]
Comment (angelazhou32) [reply asynchronous-activity-completion/how-to-async-complete-activities.md:11]: changed

File: asynchronous-activity-completion/how-to-async-complete-activities.md (hunk 2 above)
Comment (angelazhou32) [reply asynchronous-activity-completion/how-to-async-complete-activities.md:36]: fixed

File: asynchronous-activity-completion/how-to-async-complete-activities.md (hunk 3 above)
Comment (angelazhou32) [reply asynchronous-activity-completion/how-to-async-complete-activities.md:37]: fixed

File: asynchronous-activity-completion/when-to-use-async-activity-completion.md [hunk 8]
```diff
@@ -0,0 +1,14 @@
+## When to Use Asynchronous Completion
+
+When an Activity initiates a computation that is completed by an external system, there are several ways to handle the communication of the final result back to the Workflow. These methods depend on the level of control you have over the external system.
```
[the remark this answers is not in the document]
Comment (angelazhou32) [reply asynchronous-activity-completion/when-to-use-async-activity-completion.md:3]: Task is better. Fixed. Thanks

File: asynchronous-activity-completion/async-completion-vs-signals.md [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+Otherwise, if the external system can reliably be trusted to do the task and Signal back with the result, and it doesn't need to Heartbeat or receive Cancellation, then you may want to use Signals.
+
+## Ideal cases for Signals:
+
+- **Reliable External Systems**: The external system is reliable in performing the task and can signal back with the result. An advantage of this is its simplicity. The Workflow does not need to manage or monitor the ongoing process, it just waits for the final Signal indicating completion.
+- **No need for Heartbeats**: The task does not require continuous monitoring (heartbeats)
+
+One of the benefits of using Signals in Temporal Workflows relates to how the system handles retries in case of failures, especially in long-duration tasks. For example, consider an external process that is waiting for a human to review something and respond, and they could take up to a week to do so. If you use Async Completion, you would:
+
+- Set a long Start-To-Close Timeout that would align with the expected duration of the human review process (e.g., one week)
```
[the remark this answers is not in the document]
Comment (angelazhou32) [reply asynchronous-activity-completion/async-completion-vs-signals.md:21]: added

File: asynchronous-activity-completion/how-to-async-complete-activities.md [hunk 10]
```diff
@@ -0,0 +1,41 @@
+# How to Asynchronously Complete Activities
+
+Asynchronous Activity Completion calls can take a Task Token in as an argument. A Task Token is a unique identifier for an Activity Task Execution. It's essential for Temporal to track the specific Activity instance across machines.
+
+To use Asynchronous Activity Completions, there are three steps to follow:
+
+1. **Provide Identifying Information**: The Activity provides the external system with identifying information needed to complete the Activity Execution. Identifying information can be a Task Token or a combination of Namespace, Workspace ID, and Activity ID.
+
+2. **Async Activity Completion**: The Activity Function completes in a way that identifies it as waiting to be completed by an external system. You can throw [`CompleteAsyncError`](https://typescript.temporal.io/api/classes/activity.CompleteAsyncError) to make the Worker forget about this Activity. The Activity can then be completed asynchronously elsewhere.
```
[the remark this answers is not in the document]
Comment (angelazhou32) [reply asynchronous-activity-completion/how-to-async-complete-activities.md:9]: added

File: asynchronous-activity-completion/what-is-async-activity-completion.md [hunk 11]
```diff
@@ -0,0 +1,15 @@
+# What is Asynchronous Activity Completion?
+
+There are some scenarios when it is not desirable for an Activity to be completed upon the completion of a function. Imagine a case where an Activity requires user input to reach a conclusion, or perhaps it needs to wait for an external event from a third-party service.
```
[the remark this answers is not in the document]
Comment (angelazhou32) [reply asynchronous-activity-completion/what-is-async-activity-completion.md:3]: Yeah the completion of the Activity function. I rewrote this paragraph to hopefully be more clear! I agree that it was a bit hard to follow.

File: asynchronous-activity-completion/what-is-async-activity-completion.md [hunk 12]
```diff
@@ -0,0 +1,15 @@
...
+
+Remember, Activity Heartbeats act as signals that the Activity is still in progress. Asynchronous Activity Completion is a feature that enables an Activity Function to return without causing the Activity Execution to complete.  The Temporal Client can then be used from anywhere to both Heartbeat Activity Execution progress and eventually complete the Activity Execution and provide a result, providing flexibility in handling long-running or externally dependent tasks.
```
[the remark this answers is not in the document]
Comment (angelazhou32) [reply asynchronous-activity-completion/what-is-async-activity-completion.md:5]: rewritten!

File: asynchronous-activity-completion/what-is-async-activity-completion.md [hunk 13]
```diff
@@ -0,0 +1,15 @@
+# What is Asynchronous Activity Completion?
+
+In certain situations, it's not correct for an Activity to be marked as 'completed' immediately or shortly after its function executes. Consider a scenario where the Activity depends on receiving user input that might take some time, or it needs to wait awhile to process a large data set. In these cases, the Activity cannot 'complete' immedietely, because it's still awaiting important information or an event to finalize its process.
```
[the remark this answers is not in the document]
Comment (tomwheeler) [reply asynchronous-activity-completion/what-is-async-activity-completion.md:3]: Instead of this:

> In certain situations, it's not correct for an Activity to be marked as 'completed' immediately or shortly after its function executes. 

I would suggest more strongly reiterating what "normal" Activity completion is and expressing that as being undesirable in some cases (i.e., at odds with the business logic you wish to implement) rather than being incorrect (i.e., a flaw that could harm the system):

> As you've previously learned, an Activity Execution is marked as completed when its function returns a value. This behavior is not always desirable, because certain situations may require additional interaction before the Activity is regarded as complete. 


it's not desirable for an Activity to be marked as "completed" when its function executes.

File: asynchronous-activity-completion/how-to-async-complete-activities.md [hunk 14]
```diff
@@ -0,0 +1,47 @@
+# How to Asynchronously Complete Activities
+
+Asynchronous Activity Completion uses Task Tokens to track the specific Activity that was executed. Task Tokens are unique identifiers for an Activity Task Execution and are passed into the Activity as an argument. It's essential for Temporal to track the specific Activity instance across machines.
+
+To use Asynchronous Activity Completions, there are three steps to follow:
```
Comment (tomwheeler) [inline asynchronous-activity-completion/how-to-async-complete-activities.md:5]: This says "three steps to follow" but I initially saw only two. Only after I read a few paragraphs further did I see the third. 

I'd suggest keeping the list tight, with only a short summary of each step, and going into more detail about each step after the list (possibly adding a heading corresponding to each step and providing the detail below that, similar to what you did on the [Developing Signals](https://github.com/temporalio/edu-interacting-with-workflows-typescript-content/blob/signals/signaling-your-workflows/developing-signals.md) page)

File: asynchronous-activity-completion/exercise-async-activity-completion.md [hunk 15]
```diff
@@ -0,0 +1,10 @@
+# Hands-On Exercise #4: Asynchronous Activity Completion
+
+During this exercise, you will:
+
+- Retrieve a task token from your Activity execution
+- Learn to structure asynchronous Activities
```
Comment (tomwheeler) [inline asynchronous-activity-completion/exercise-async-activity-completion.md:6]: I would change this to "Structure asynchronous Activities" since they presumably learned this earlier and are now practicing what they learned. 

OTOH, that does not sound like something they actually do in this exercise (according to the instructions). Maybe it's better to delete this item. Whatever you decide to do, be sure to also make the same change to the README for this exercise in the code repo.

**Review by tomwheeler — CHANGES_REQUESTED** [review]

File: asynchronous-activity-completion/README.md [hunk 16]
```diff
@@ -0,0 +1,22 @@
...
+
+- Explain Asynchronous Activity Completion and how it is used to provide flexibility in handling long-running or externally dependent tasks.
+- Identify use cases and scenarios where a developer would use Asynchronous Activity Completion as opposed to Signals, and the advantages of both
+- Retrieve a task token from your Activity execution
+- Asynchronously complete an Activity with the SDK
+
+### 💻 Lessons
+
+- [What is Asynchronous Activity Completion?](what-is-async-activity-completion.md)
+- [When to Use Asynchronous Activity Competion](when-to-use-async-activity.md)
```
[the remark this answers is not in the document]
Comment (angelazhou32) [reply asynchronous-activity-completion/README.md:19]: fixed

File: asynchronous-activity-completion/README.md [hunk 17]
```diff
@@ -0,0 +1,22 @@
...
+- [Asynchronous Activity Competion vs. Signals](async-completion-vs-signals)
```
[the remark this answers is not in the document]
Comment (angelazhou32) [reply asynchronous-activity-completion/README.md:20]: fixed

File: asynchronous-activity-completion/async-completion-vs-signals.md [hunk 18]
```diff
@@ -0,0 +1,54 @@
...
+- **No Need for Heartbeats**: The task does not require continuous monitoring (heartbeats)
+
+## Consider How Often Failure is Retried
+
+Asynchronous Activities typically require a long Start-To-Close timeout when they are required to run for an extended period. When you have an Activity that can take a long time to complete, such as processes that depend on human interaction (e.g., manual reviews or approvals), the Start-To-Close timeout needs to be set long enough to account for the expected duration of these tasks. This is important because if the timeout is too short, the Activity might be prematurely marked as failed and retried or terminated, even though it is proceeding correctly but just taking a long time.
+
+Setting a long Start-To-Close timeout has trade-offs. While it allows long-running Activvities to complete, it also means that in case of actual failures, the system might take longer to recognize and address these issues, as it would wait for the timeout to expire before retrying.
```
[the remark this answers is not in the document]
Comment (angelazhou32) [reply asynchronous-activity-completion/async-completion-vs-signals.md:23]: fixed

File: asynchronous-activity-completion/async-completion-vs-signals.md [hunk 19]
```diff
@@ -0,0 +1,54 @@
...
+
+One of the benefits of using Signals in Temporal Workflows relates to how the system handles retries in case of failures, especially in long-duration tasks. For example, consider an external process that is waiting for a human to review something and respond, and they could take up to a week to do so. If you use Async Completion, you would:
+
+- Set a long Start-To-Close Timeout that would align with the expected duration of the human review process (e.g., one week).
+- If the Activity (e.g., notifying the external system) fails early and doesn't report an error, it won't be retried until the timeout is hit (a week later), leading to a delayed response.
+
+In contrast, if you use Signals, you would:
+
+- Set a short Start-To-Close Timeout (e.g., one minute)
+- If the activity fails early in the process, it will be retried much sooner (in a minute), allowing quicker resolution of the failure.
```
[the remark this answers is not in the document]
Comment (angelazhou32) [reply asynchronous-activity-completion/async-completion-vs-signals.md:33]: fixed

File: asynchronous-activity-completion/async-completion-vs-signals.md [hunk 20]
```diff
@@ -0,0 +1,54 @@
...
+
+In the second scenario, the failure is retried sooner. This is particularly helpful in scenarios like this in which the external process might take a long time.
+
+## Example Scenario
+
+Let's look at an example scenario of a user sign-up Workflow and explore what steps could be used with a Signal vs. Asynchronous Activity Completion:
```
[the remark this answers is not in the document]
Comment (angelazhou32) [reply asynchronous-activity-completion/async-completion-vs-signals.md:39]: fixed

File: asynchronous-activity-completion/async-completion-vs-signals.md [hunk 21]
```diff
@@ -0,0 +1,54 @@
...
+1. Upon receiving the user e-mail address, Temporal starts a `createSendToken` Activity: _Async. Activity Completion_
+2. User receives an e-mail with a verification token link: _Async. Activity Completion_
+3. User clicks verification token link: _Signal_
+4. User enters phone number: _Signal_
+5. User receives one-time password with SMS: _Async. Activity Completion_
+6. User enters one-time password: _Signal_
+
+While you could technically use Signals for all of these steps, Asynchronous Activity Completion are a better fit for some steps.
+
+Asynchronous activities are ideal for tasks that are operational and system-driven, such as sending an email or SMS. These tasks are generally automated and do not require direct, manual input from a user at the moment of execution. These Activities often need robust error handling. For instance, if an email fails to send due to a server issue, the Activity can be retried automatically based on predefined retry policies. Asynchronous activities in Temporal are designed to handle such retries and timeouts effectively.
```
[the remark this answers is not in the document]
Comment (angelazhou32) [reply asynchronous-activity-completion/async-completion-vs-signals.md:50]: Fixed

File: asynchronous-activity-completion/async-completion-vs-signals.md [hunk 22]
```diff
@@ -0,0 +1,54 @@
...
+
+Manual activities, such as a user clicking a link or entering data, occur at an unpredictable pace. These are dependent on the user's actions, which can vary greatly in timing. Signals are designed for scenarios where the Workflow needs to respond to external, asynchronous events. When a user performs an action, a Signal is sent to the Workflow, notifying it of the event without the Workflow having to poll or wait actively for the action. With Signals, a developer can use a very short running Activity (with a very brief `StartToCloseTimeout`) that makes a request to a third-party service and then waits for the Activity to complete. Once the Activity completes, a Signal will notify the application that it is ready for the next step.
+
+Another reason for using Signals for manual interactions is that such manual interactions frequently have multiple states. For example, a task can be available, assigned, timed out, completed, etc. Signals provide a flexible way to model these external events and track state transitions within a Workflow.
```
[the remark this answers is not in the document]
Comment (angelazhou32) [reply asynchronous-activity-completion/async-completion-vs-signals.md:54]: Fixed

File: asynchronous-activity-completion/how-to-async-complete-activities.md [hunk 23]
```diff
@@ -0,0 +1,47 @@
...
+
+1. **Provide Identifying Information**: The Activity provides the external system with identifying information needed to complete the Activity Execution. Identifying information can be a Task Token or a combination of Namespace, Workspace ID, and Activity ID.
```
[the remark this answers is not in the document]
Comment (angelazhou32) [reply asynchronous-activity-completion/how-to-async-complete-activities.md:7]: Whoops! Fixed

File: asynchronous-activity-completion/what-is-async-activity-completion.md (hunk 13 above)
[the remark this answers is not in the document]
Comment (angelazhou32) [reply asynchronous-activity-completion/what-is-async-activity-completion.md:3]: thanks! used the suggestion

File: asynchronous-activity-completion/what-is-async-activity-completion.md [hunk 24]
```diff
@@ -0,0 +1,15 @@
...
+
+Asynchronous Activity Completion addresses this need. This is a feature that allows an asynchronous Activity function to execute and return, but without marking the Activity Execution as complete. Instead, the Activity remains 'open' or 'in-progress', even after the function has ended.
```
[the remark this answers is not in the document]
Comment (angelazhou32) [reply asynchronous-activity-completion/what-is-async-activity-completion.md:5]: done

File: asynchronous-activity-completion/when-to-use-async-activity-completion.md [hunk 25]
```diff
@@ -0,0 +1,14 @@
...
+
+1. **Asynchronous Completion by the External System**: The external system directly informs Temporal when it has completed the task using Asynchronous Completion. This means the Activity in the Workflow starts the process, but the completion and the result are reported back by the external system. This method is ideal for long-running processes, since Heartbeats will be sent to indicate the ongoing Activity. It's also important for scenarios where you can monitor the progress and ensure the Activity is alive.
+2. **Signal from External System**: The Activity completes normally with the final result. Once the result is ready, the external system sends a Signal to the Workflow with the result.
+3. **Polling by Subsequent Activity**: A subsequent Activity polls the external system for the result.
+
+In the next chapter, we will delve deeper into when a developer might choose to use a Signal and when to choose to use Asynchronous Completion.
```
[the remark this answers is not in the document]
Comment (angelazhou32) [reply asynchronous-activity-completion/when-to-use-async-activity-completion.md:9]: I meant 'in the next section' which is async activity completion vs. signals! I will switch it to say that.

File: asynchronous-activity-completion/what-is-async-activity-completion.md [hunk 26]
```diff
@@ -0,0 +1,15 @@
+# What is Asynchronous Activity Completion?
+
+As you've previously learned, an Activity Execution is marked as completed when its function returns a value. This behavior is not always desirable, because certain situations may require additional interaction before the Activity is regarded as complete. Consider a scenario where the Activity depends on receiving user input that might take some time, or it needs to wait awhile to process a large data set. In these cases, the Activity cannot "complete" immediately, because it's still awaiting important information or an event to finalize its process.
```
Comment (mjameswh) [inline asynchronous-activity-completion/what-is-async-activity-completion.md:3]: Great explanation, but I think the example given below isn't strong enough to emphasis the pertinence of Async Completion. I'd suggest instead to consider the case where the activity execution gets delegated to a different process (even on a different machine), which the activity can't monitor directly.

For example, imagine that my activity starts a Lambda to handle the block. From that point on, its the code on that lambda that will handle execution of the activity, so my initial activity worker can simply go away.

File: asynchronous-activity-completion/how-to-async-complete-activities.md [hunk 27]
```diff
@@ -0,0 +1,53 @@
+# How to Asynchronously Complete Activities
+
+Asynchronous Activity Completion uses Task Tokens to track the specific Activity that was executed. Task Tokens are unique identifiers for an Activity Task Execution and are passed into the Activity as an argument. It's essential for Temporal to track the specific Activity instance across machines.
+
+To use Asynchronous Activity Completions, there are three steps to follow:
+
+1. Provide the external system identifying information of the Activity.
+2. Throw a `CompleteAsyncError` Error to indicate to the Worker that Activity is not yet complete but to "forget" about the Activity momentarily.
+3. Optionally, while the external work is being done, the Temporal Client can send Heartbeats to update the progress of the Activity Execution.
```
Comment (mjameswh) [inline asynchronous-activity-completion/how-to-async-complete-activities.md:9]: There ought to be a fourth step: Actually complete the Activity, using the Async Completion Client.

Comment (mjameswh) [inline asynchronous-activity-completion/how-to-async-complete-activities.md:9]:
```suggestion
3. Optionally, while the external work is being done, the Temporal Async Completion Client can be used to send Heartbeats and provide update on progression of the Activity Execution.
```

File: asynchronous-activity-completion/how-to-async-complete-activities.md [hunk 28]
```diff
@@ -0,0 +1,53 @@
...
+
+Let's dive into these above steps in more detail.
+
+1. **Provide Identifying Information**: The Activity provides the external system with identifying information needed to complete the Activity Execution. Identifying information can be a Task Token or a combination of Namespace, Workflow ID, and Activity ID.
+
+2. **Async Activity Completion**: When an Activity is supposed to be completed by an external system, the Activity function signals this by throwing a [`CompleteAsyncError`](https://typescript.temporal.io/api/classes/activity.CompleteAsyncError). This indicates that the Activity is not yet complete.
+
+When the `CompleteAsyncError` is thrown, the Worker that is executing the Activity understands that this Activity is not yet complete and that it should not wait for the Activity to finish in the normal synchronous manner.
```
Comment (mjameswh) [inline asynchronous-activity-completion/how-to-async-complete-activities.md:17]: Indentation for these paragraphs are incorrect. Don't know if our LMS's markdown implementation support this, but generally, you would need to add 4 spaces at the beginning of this paragraph so that it aligns properly with the numbered paragraph above.

File: asynchronous-activity-completion/how-to-async-complete-activities.md [hunk 29]
```diff
@@ -0,0 +1,53 @@
...
+
+To use Asynchronous Activity Completions, there are three steps to follow:
+
+1. Provide the external system identifying information of the Activity.
+2. Throw a `CompleteAsyncError` Error to indicate to the Worker that Activity is not yet complete but to "forget" about the Activity momentarily.
+3. Optionally, while the external work is being done, the Temporal Client can send Heartbeats to update the progress of the Activity Execution.
+
+Let's dive into these above steps in more detail.
+
+1. **Provide Identifying Information**: The Activity provides the external system with identifying information needed to complete the Activity Execution. Identifying information can be a Task Token or a combination of Namespace, Workflow ID, and Activity ID.
```
Comment (mjameswh) [inline asynchronous-activity-completion/how-to-async-complete-activities.md:13]: I think we haven't documented this anywhere yet, but in cases where an Activity Task Token is shared with an external party, it should be treated as security-sensitive information. That is, they should encrypt a task token before sharing it with external parties, and should verify authenticity of a task token when receiving such a token.

For context: even though they appear to be opaque tokens, an Activity Task Token really are protobuf serialized structures that contains things like the workflow id and run id, the activity's type, position in history, etc. Therefore, exposing an Activity Task Token to third parties could leak some internal details about the workflow. It would also be possible for someone to maliciously-alter the content of a token before passing it back to the application. The probability of success of such an attack seems very low, but it's still pertinent to encourage devs to be careful.

File: asynchronous-activity-completion/how-to-async-complete-activities.md (hunk 28 above)
Comment (mjameswh) [inline asynchronous-activity-completion/how-to-async-complete-activities.md:17]: > When the `CompleteAsyncError` is thrown, the Worker that is executing the Activity understands that this Activity is not yet complete and that it should not wait for the Activity to finish in the normal synchronous manner.
>
> Temporal marks this Activity as 'in progress' but does not actively track it for completion within the Worker. This essentially makes the Worker "forget" about the Activity.

I'm slightly uncomfortable with the language we use to describe this process. This makes it sounds like if this was some kind of magic trick that you play on the worker. I mean, Async Task Completion really goes back to the cluster.

I suggest something like this:

By throwing `CompleteAsyncError`, the Activity indicates to the Temporal Cluster that further progress regarding this activity (including completion or failure, as well as periodic heartbeating) will come in from a different system. That also means that this Activity Worker is no longer responsible for this activity task, and can therefore poll new activity tasks.

File: asynchronous-activity-completion/how-to-async-complete-activities.md [hunk 30]
```diff
@@ -0,0 +1,53 @@
...
+
+When the `CompleteAsyncError` is thrown, the Worker that is executing the Activity understands that this Activity is not yet complete and that it should not wait for the Activity to finish in the normal synchronous manner.
+
+Temporal marks this Activity as 'in progress' but does not actively track it for completion within the Worker. This essentially makes the Worker "forget" about the Activity.
+
+The actual completion of the Activity is then handed over to an external system. This external system is responsible for eventually signaling the completion of the Activity.
+
+3. **Use Temporal Client**: Optionally, while the external work is being done, the Temporal Client can send Hearbeats to update the progress of the Activity Execution. Once the completion signal is received from the external system, Temporal then resumes the Workflow execution from the point where the Activity was left off. To asynchronously complete an Activity, call[`AsyncCompletionClient`](https://typescript.temporal.io/api/classes/client.AsyncCompletionClient) which is a client for asynchronous completion and Heartbeating of Activities.
```
Comment (mjameswh) [inline asynchronous-activity-completion/how-to-async-complete-activities.md:23]:
```suggestion
3. **Heartbeating**: While working on the activity, the external system may use an [`AsyncCompletionClient`](https://typescript.temporal.io/api/classes/client.AsyncCompletionClient) to send Activity Hearbeats to the cluster. Doing so allows the activity to provide progress details regarding the Activity Execution. It also informs the cluster that the activity is still actively being worked on. Failure to send heartbeats before the expiration of `heartbeatTimeout` will cause the server to retry the activity.

4. **Completing the activity**: When appropriate, the external system notifies the Temporal Cluster that the activity has completed by using an [`AsyncCompletionClient`](https://typescript.temporal.io/api/classes/client.AsyncCompletionClient).

    Once the completion signal is received from the external system, Temporal then resumes the Workflow execution from the point where the Activity was left off. 
```

File: asynchronous-activity-completion/how-to-async-complete-activities.md [hunk 31]
```diff
@@ -0,0 +1,53 @@
...
+
+```typescript
+import { CompleteAsyncError, Context } from "@temporalio/activity";
+import { AsyncCompletionClient } from "@temporalio/client";
+
+export async function doSomethingAsync(): Promise<string> {
+  //extracts task token
+  const taskToken = Context.current().info.taskToken;
```
Comment (mjameswh) [inline asynchronous-activity-completion/how-to-async-complete-activities.md:31]:
```suggestion
  const taskToken = activityInfo().taskToken;
```

File: asynchronous-activity-completion/how-to-async-complete-activities.md [hunk 32]
```diff
@@ -0,0 +1,53 @@
...
+
+Temporal marks this Activity as 'in progress' but does not actively track it for completion within the Worker. This essentially makes the Worker "forget" about the Activity.
+
+The actual completion of the Activity is then handed over to an external system. This external system is responsible for eventually signaling the completion of the Activity.
+
+3. **Use Temporal Client**: Optionally, while the external work is being done, the Temporal Client can send Hearbeats to update the progress of the Activity Execution. Once the completion signal is received from the external system, Temporal then resumes the Workflow execution from the point where the Activity was left off. To asynchronously complete an Activity, call[`AsyncCompletionClient`](https://typescript.temporal.io/api/classes/client.AsyncCompletionClient) which is a client for asynchronous completion and Heartbeating of Activities.
+
+```typescript
+import { CompleteAsyncError, Context } from "@temporalio/activity";
```
Comment (mjameswh) [inline asynchronous-activity-completion/how-to-async-complete-activities.md:26]:
```suggestion
import { CompleteAsyncError, activityInfo } from "@temporalio/activity";
```

File: asynchronous-activity-completion/how-to-async-complete-activities.md [hunk 33]
```diff
@@ -0,0 +1,53 @@
...
+  // does some work...
+  await client.complete(taskToken, "Job's done!");
+}
+```
+
+Let's look at the above code step by step:
+
+1. We began by extracting the Task Token that uniquely identifies the current Activity execution.
+2. We then scheduled `doSomeWork` to be executed after a delay. The delay represents a stand-in for more complex asynchronous operations, like waiting for an external API response or processing that happens in a separate system.
+3. We then threw the `CompleteAsyncError` to make the Worker forget about this Activity. This signaled that the Activity has not completed yet and is waiting for an external completion, not immediately upon the function's return.
```
Comment (mjameswh) [inline asynchronous-activity-completion/how-to-async-complete-activities.md:48]:
```suggestion
3. We then threw the `CompleteAsyncError` to inform the Cluster that this Activity's completion will come through a different path (i.e., the current Worker is no longer responsible for this Activity Task).
```

**Review by mjameswh — CHANGES_REQUESTED** [review]
See review comments.
